File: medium/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  // O(mn) space
  const map = new Map();

  // O(m) time
  strs.forEach((str) => {
    // O(nlogn) time, sorting is bottleneck
    const sorted = Array.from(str).sort().join('');

    if (!map.has(sorted)) {
      map.set(sorted, []);
    }

    map.get(sorted).push(str);
  });

  // O(mn) space
  return Array.from(map.values());
};

// m is the length of strs, n is the length of the longest str in strs
// Time complexity: O(mnlogn)
// Space complexity: O(mn)

File: medium/0049-group-anagrams.py
from typing import List


# If n is the number of strings in strs and m is the longest string:
# Time complexity: O(n * mlog(m))
# Space complexity: O(n * m)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        d = {}
        for str in strs:
            sortedStr = "".join(sorted(str))
            if sortedStr in d:
                d[sortedStr].append(str)
            else:
                d[sortedStr] = [str]

        return list(d.values())
